### PR TITLE
task 154, task 152 of US139 convey the quality of task process

### PR DESCRIPTION
**Pull request by xiangweizheng**
1 Create a API call. This call assess a task based on task_assessment_state_trans based on task Id.

Test code: add following code in Taiga.js file
import {task_assessment_state_trans}  from '../libraries/Taiga'

{console.log(task_assessment_state_trans(2577741).then((val) => {console.log('fe', val)} ))}

Expected result:
fe Array(3) 0: {state_trans_invalid: false, date: 1541629794485, user: {…}, status_trans: Array(2)} 1: {state_trans_invalid: false, date: 1542392039286, user: {…}, status_trans: Array(2)} 2: {state_trans_invalid: false, date: 1543257817335, user: {…}, status_trans: Array(2)} length: 3 __proto__: Array(0)

2 Create a call return a task current end status and estimated finished percentage based on task Id

Test code: add following code in Taiga.js file
{console.log(task_assessment_endstate(2577741).then((val) => {console.log('fe', val)} )
import {task_assessment_endstate}  from '../libraries/Taiga'

expected result:
fe {finished: true, end_status: "Closed", num_stat: 3, fin_per: 100}

### CODE REVIEW

**Comment by PaHorton** [comment]
I am not on the taiga team so I will be removing myself as a reviewer

**Comment by HalcyonAura** [comment]
I have mentioned this is the past, but please refrain from having multiple tasks completed in a single PR.

File: taigit/src/libraries/Taiga.ts [hunk 1]
```diff
@@ -78,4 +78,124 @@ task_history(taskId : number) : Promise<Object> {
...
+        &&(new_entry.status_trans[0]!="Ready for test"&&new_entry.status_trans[1]!="Closed"))
+        new_entry.state_trans_invalid= true;
+    output.push(new_entry);
+}
+
+return output;
+}
+
+
+**
```
Comment (HalcyonAura) [inline taigit/src/libraries/Taiga.ts]: static analysis: missing leading /

File: taigit/src/libraries/Taiga.ts [hunk 2]
```diff
@@ -78,4 +78,124 @@ task_history(taskId : number) : Promise<Object> {
...
+let output : Array<Object> = [];
+for(let entry of data) {
+    let new_entry = {
+        state_trans_invalid: false,
+        date : new Date(entry.created_at).getTime(),
+        user : entry.user,
+        status_trans:entry.values_diff.status
+    }
+
+    //Only three status transistion is valid
```
Comment (HalcyonAura) [inline taigit/src/libraries/Taiga.ts]: From an initial review, I think you should consider the instances where students make incorrect transitions. There are times where items are moved by accident (ie things are fixed within a certain timeframe), or that tasks must be returned to certain stages to accommodate things.
This analysis was intended to not only provide a way to assess a student's process (you seem to have accomplished this), but also show the progression of a task for frontend to display. I am not sure your current code allows for this (this comment is not meant to change your development but rather to inspire more considerations for your work)

File: taigit/src/libraries/Taiga.ts [hunk 3]
```diff
@@ -78,4 +78,124 @@ task_history(taskId : number) : Promise<Object> {
...
+    let output : Array<Object> = [];
+    let endstatus:string = "New";
+    for(let entry of data) {
+        let new_entry = {
+            state_trans_invalid: false,
+            date : new Date(entry.created_at).getTime(),
+            user : entry.user,
+            status_trans:entry.values_diff.status
+        }
+        endstatus=new_entry.status_trans[1];
```
Comment (HalcyonAura) [inline taigit/src/libraries/Taiga.ts]: static analysis: you have inconsistent variable name styles. please pick a style (or review previous methods to see what the trend is) and stick with it. If there is inconsistency across all functions please alert me of this so I can address this in my documentation task for earlier functions

File: taigit/src/libraries/Taiga.ts [hunk 4]
```diff
@@ -78,4 +78,124 @@ task_history(taskId : number) : Promise<Object> {
...
+            t_fin_per = 60;
+            t_num_stat = 2;
+            break
+        }
+        case "In progress":{
+            t_fin_per = 10;
+            t_num_stat = 1;
+            break
+        }
+        case "New":{
```
Comment (HalcyonAura) [inline taigit/src/libraries/Taiga.ts]: why does a "new" case have a higher finished % (I think? Your variable names are a little hard to follow) than in progress?
Do you have research to back up these percentage distributions? (It's okay if you don't right now, I just think we should not have magic numbers dictating analysis rather some real world backing on how to assess this so please consider this for the future of your work)

Comment (HalcyonAura) [reply taigit/src/libraries/Taiga.ts]: to build on this. I don't understand the correlation between t_fin_per and t_num_stat nor understand how this will convey value to a potential user. I feel a comment explaining this should exist in the code whether to a link or a brief summary. If there is no reasonable correlation then there should be work to synthesize one.

File: taigit/src/libraries/Taiga.ts [hunk 5]
```diff
@@ -78,4 +78,124 @@ task_history(taskId : number) : Promise<Object> {
...
-}
\ No newline at end of file
+}
+
+/**
+ * @summary This call assess a task based on task_assessment_state_trans based on task Id
+ * @param taskId the ID for the task to assess task
+ * @returns  whether the task's state transition is valid,task status transition, date based on task Id
+ * * {
+ * Current in Use
```
Comment (HalcyonAura) [inline taigit/src/libraries/Taiga.ts]: I do not believe your current in use/can be added notation are valid javadocs and am inclined to believe those are notes left to yourself?

**Comment by xiangweizheng** [comment]
> I have mentioned this is the past, but please refrain from having multiple tasks completed in a single PR.
I can make seperate pull request but I think I delay the pull request even further..

File: taigit/src/libraries/Taiga.ts (hunk 4 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: fix new case, that is a typo.   t_num_stat  will translate state "New", "In progress",etc into number so I can be easily plotted.  percentage distributions is used to have a inspect of  a sprint while it is going on, and provide a additional view of burndown chart. The numbers I picked is from statistical data. For users using our tool, they can change the estimated percentage based on their experience.

File: taigit/src/libraries/Taiga.ts (hunk 3 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: fixed.

File: taigit/src/libraries/Taiga.ts (hunk 2 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: I think about the situation when I write the API. It can be an improvement in the future sprint.

File: taigit/src/libraries/Taiga.ts [hunk 6]
```diff
@@ -78,4 +78,118 @@ task_history(taskId : number) : Promise<Object> {
...
+        date : new Date(entry.created_at).getTime(),
+        user : entry.user,
+        status_trans:entry.values_diff.status
+    }
+
+    //Only three status transistion is valid
+    //which is  ["New", "In progress"] ["In progress", "Ready for test"]  ["Ready for test", "Closed"])
+    if((new_entry.status_trans[0] != "New"&&new_entry.status_trans[1] != "In progress")
+        &&(new_entry.status_trans[0] != "In progress"&&new_entry.status_trans[1] != "Ready for test")
+        &&(new_entry.status_trans[0] != "Ready for test"&&new_entry.status_trans[1] != "Closed"))
```
Comment (zawata) [inline taigit/src/libraries/Taiga.ts]: this logic is really hard to follow and im not sure it would work. 
How about: 

```js
    let old_status = new_entry.status_trans[0];
    let new_status = new_entry.status_trans[1];
    if(     (old_status == "New"            && new_status != "In progress") ||
            (old_status == "In progress"    && new_status != "Ready for test") ||
            (old_status == "Ready for test" && new_status != "Closed"))
```

File: taigit/src/libraries/Taiga.ts [hunk 7]
```diff
@@ -78,4 +78,118 @@ task_history(taskId : number) : Promise<Object> {
...
+ *      fin_per : number            // finished percentage % default [0%,10%,60%,100%]
+ *      end_status : string,       // current task status  ["New", "In progress","Ready for test", "Closed"]
+ *      num_stat : number           // current task status [0 1 2 3] corresponding to string in end_status
+ *
+ * }
+ */
+export async function
+task_assessment_endstate(taskId : number) : Promise<Object> {
+
+    let data = (await axios.get(`https://api.taiga.io/api/v1/history/task/${taskId}`)).data;
```
Comment (zawata) [inline taigit/src/libraries/Taiga.ts]: we already have a task history function. why repeat the call?

File: taigit/src/libraries/Taiga.ts [hunk 8]
```diff
@@ -78,4 +78,118 @@ task_history(taskId : number) : Promise<Object> {
...
+
+    let t_finished : boolean;
+    let t_fin_per : number;
+    let t_num_stat : number;
+     if(endstatus != "Closed") {
+         t_finished = false;
+     }
+    else{
+        t_finished = true;
+    }
```
Comment (zawata) [inline taigit/src/libraries/Taiga.ts]: `let t_finished : boolean = (endstatus == "Closed")`?

File: taigit/src/libraries/Taiga.ts (hunk 6 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: If everyone think it is good, it is fine by me. I can change this.

File: taigit/src/libraries/Taiga.ts (hunk 8 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: OK. If it is the same logic, it is fine.

File: taigit/src/libraries/Taiga.ts (hunk 7 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: I do not know the other way to get the information I required.

File: taigit/src/libraries/Taiga.ts (hunk 8 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: replace this line with your comment, it seems a error/ or I did not get it right. I will go with the conventional way to prevent error.

File: taigit/src/libraries/Taiga.ts (hunk 6 above)
Comment (zawata) [reply taigit/src/libraries/Taiga.ts]: I didn't test it with your PR, its just a thought you'll have to make sure it works

File: taigit/src/libraries/Taiga.ts (hunk 7 above)
Comment (zawata) [reply taigit/src/libraries/Taiga.ts]: I made a function called `task_history()` that does exactly this call.

File: taigit/src/libraries/Taiga.ts (hunk 6 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: change it the way as you provided.

File: taigit/src/libraries/Taiga.ts (hunk 7 above)
Comment (xiangweizheng) [reply taigit/src/libraries/Taiga.ts]: However, I get the data from the task history, and provide further processing with the data.

**Comment by HalcyonAura** [comment]
> code works but im recommending you modify it to use the task_history function.

I believe this may be the better route or even to expect frontend to pass in the data returned from the task_history function as a parameter to this work. As we progress with more analysis we'll want to reduce the amount of duplicate/excess API calls to ensure frontend can retrieve data slightly faster from us. At least, that is the preference that has been conveyed from frontend in our general chat if you would like more concrete evidence of this.

**Comment by xiangweizheng** [comment]
> > code works but im recommending you modify it to use the task_history function.
> 
> I believe this may be the better route or even to expect frontend to pass in the data returned from the task_history function as a parameter to this work. As we progress with more analysis we'll want to reduce the amount of duplicate/excess API calls to ensure frontend can retrieve data slightly faster from us. At least, that is the preference that has been conveyed from frontend in our general chat if you would like more concrete evidence of this.

As current sprint is almost finished, this API is actually working. As you may say it is slight slower. I could change it to use task_history in the next sprint. Currently, I do not know how to use task_history in my API. I have a try to change to use task_history but reports error.

**Comment by xiangweizheng** [comment]
> > code works but im recommending you modify it to use the task_history function.
> 
> I believe this may be the better route or even to expect frontend to pass in the data returned from the task_history function as a parameter to this work. As we progress with more analysis we'll want to reduce the amount of duplicate/excess API calls to ensure frontend can retrieve data slightly faster from us. At least, that is the preference that has been conveyed from frontend in our general chat if you would like more concrete evidence of this.

Talk with john and work out a fix for the error.

**Comment by xiangweizheng** [comment]
> > code works but im recommending you modify it to use the task_history function.
> 
> I believe this may be the better route or even to expect frontend to pass in the data returned from the task_history function as a parameter to this work. As we progress with more analysis we'll want to reduce the amount of duplicate/excess API calls to ensure frontend can retrieve data slightly faster from us. At least, that is the preference that has been conveyed from frontend in our general chat if you would like more concrete evidence of this.

So spend hours to change into use" taks_history" but it still have errors. The problem happen when task_history output  json object from Taiga, it simply treat it as an Object. The object  do not have all the well-edited interface as json. You need to define the object once you use it(once you dive into the json object). It is way too inconvenient to program once you dive into the json object. I will keep the working version right now.  I do not think the slightly faster on the frontend worth hours working on this. Also there is no hard requirement of the speed.

**Comment by HalcyonAura** [comment]
Closing this PR as it is obsolete.
